File: app/src/main/java/xyz/klinker/messenger/fragment/settings/MmsConfigurationFragment.java
package xyz.klinker.messenger.fragment.settings;

import android.Manifest;
import android.app.AlertDialog;
import android.content.DialogInterface;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.preference.PreferenceFragment;
import android.support.v4.content.ContextCompat;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;

import com.codekidlabs.storagechooser.StorageChooser;

import xyz.klinker.messenger.R;
import xyz.klinker.messenger.api.implementation.Account;
import xyz.klinker.messenger.api.implementation.ApiUtils;
import xyz.klinker.messenger.shared.data.MmsSettings;
import xyz.klinker.messenger.shared.data.Settings;

public class MmsConfigurationFragment extends PreferenceFragment {

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        addPreferencesFromResource(R.xml.settings_mms);
        initConvertToMMS();
        initMaxImageSize();
        initGroupMMS();
        initAutoSaveMedia();
        initDownloadLocation();

        // MMS APN settings
        initOverrideSystemSettings();
        initMmsc();
        initProxy();
        initPort();
        initUserAgent();
        initUserAgentProfileUrl();
        initUserAgentProfileTagName();
    }

    @Override
    public void onStop() {
        super.onStop();
        Settings.get(getActivity()).forceUpdate();
    }


    private void initConvertToMMS() {
        findPreference(getString(R.string.pref_convert_to_mms))
                .setOnPreferenceChangeListener((preference, o) -> {
                    String convert = (String) o;
                    new ApiUtils().updateConvertToMMS(Account.get(getActivity()).accountId,
                            convert);
                    return true;
                });
    }

    private void initMaxImageSize() {
        findPreference(getString(R.string.pref_mms_size))
                .setOnPreferenceChangeListener((preference, o) -> {
                    String size = (String) o;
                    new ApiUtils().updateMmsSize(Account.get(getActivity()).accountId,
                            size);
                    return true;
                });
    }

    private void initGroupMMS() {
        findPreference(getString(R.string.pref_group_mms))
                .setOnPreferenceChangeListener((preference, o) -> {
                    boolean group = (boolean) o;
                    new ApiUtils().updateGroupMMS(Account.get(getActivity()).accountId,
                            group);
                    return true;
                });
    }

    private void initAutoSaveMedia() {
        findPreference(getString(R.string.pref_auto_save_media))
                .setOnPreferenceChangeListener((preference, o) -> {
                    boolean save = (boolean) o;
                    new ApiUtils().updateAutoSaveMedia(Account.get(getActivity()).accountId,
                            save);
                    return true;
                });
    }

    private void initDownloadLocation() {
        findPreference(getString(R.string.pref_mms_save_location))
                .setOnPreferenceClickListener(preference -> {

                    if (ContextCompat.checkSelfPermission(getActivity(), Manifest.permission.WRITE_EXTERNAL_STORAGE)
                            == PackageManager.PERMISSION_GRANTED) {
                        showStorageChooser();
                    } else {
                        if (getActivity() != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                            getActivity().requestPermissions(new String[]{Manifest.permission.WRITE_EXTERNAL_STORAGE}, -1);
                        } else {
                            showStorageChooser();
                        }
                    }

                    return false;
                });
    }

    private void showStorageChooser() {
        StorageChooser chooser = new StorageChooser.Builder()
                .withActivity(getActivity())
                .withFragmentManager(((AppCompatActivity) getActivity()).getSupportFragmentManager())
                .allowCustomPath(true)
                .setType(StorageChooser.DIRECTORY_CHOOSER)
                .actionSave(true)
                .withPreference(MmsSettings.get(getActivity()).getSharedPrefs(getActivity()))
                .build();

        chooser.show();
        chooser.setOnSelectListener(s -> { });
    }

    private void initOverrideSystemSettings() {
        findPreference(getString(R.string.pref_override_system_apn))
                .setOnPreferenceChangeListener((preference, o) -> {
                    boolean override = (boolean) o;
                    new ApiUtils().updateOverrideSystemApn(Account.get(getActivity()).accountId,
                            override);

                    if (override) {
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !android.provider.Settings.System.canWrite(getActivity())) {
                            new AlertDialog.Builder(getActivity())
                                    .setMessage(com.klinker.android.send_message.R.string.write_settings_permission)
                                    .setPositiveButton(com.klinker.android.send_message.R.string.ok, new DialogInterface.OnClickListener() {
                                        @Override
                                        public void onClick(DialogInterface dialog, int which) {
                                            Intent intent = new Intent(android.provider.Settings.ACTION_MANAGE_WRITE_SETTINGS);
                                            intent.setData(Uri.parse("package:" + getActivity().getPackageName()));
                                            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);

                                            try {
                                                startActivity(intent);
                                            } catch (Exception e) {
                                                Log.e("MainActivity", "error starting permission intent", e);
                                            }
                                        }
                                    })
                                    .show();
                        }
                    }

                    return true;
                });
    }

    private void initMmsc() {
        findPreference(getString(R.string.pref_mmsc_url))
                .setOnPreferenceChangeListener((preference, o) -> {
                    String mmsc = (String) o;
                    new ApiUtils().updateMmscUrl(Account.get(getActivity()).accountId,
                            mmsc);
                    return true;
                });
    }

    private void initProxy() {
        findPreference(getString(R.string.pref_mms_proxy))
                .setOnPreferenceChangeListener((preference, o) -> {
                    String proxy = (String) o;
                    new ApiUtils().updateMmsProxy(Account.get(getActivity()).accountId,
                            proxy);
                    return true;
                });
    }

    private void initPort() {
        findPreference(getString(R.string.pref_mms_port))
                .setOnPreferenceChangeListener((preference, o) -> {
                    String port = (String) o;
                    new ApiUtils().updateMmsPort(Account.get(getActivity()).accountId,
                            port);
                    return true;
                });
    }

    private void initUserAgent() {
        findPreference(getString(R.string.pref_user_agent))
                .setOnPreferenceChangeListener((preference, o) -> {
                    String userAgent = (String) o;
                    new ApiUtils().updateUserAgent(Account.get(getActivity()).accountId,
                            userAgent);
                    return true;
                });
    }

    private void initUserAgentProfileUrl() {
        findPreference(getString(R.string.pref_user_agent_profile_url))
                .setOnPreferenceChangeListener((preference, o) -> {
                    String uaProfileUrl = (String) o;
                    new ApiUtils().updateUserAgentProfileUrl(Account.get(getActivity()).accountId,
                            uaProfileUrl);
                    return true;
                });
    }

    private void initUserAgentProfileTagName() {
        findPreference(getString(R.string.pref_user_agent_profile_tag))
                .setOnPreferenceChangeListener((preference, o) -> {
                    String uaProfileTagName = (String) o;
                    new ApiUtils().updateUserAgentProfileTagName(Account.get(getActivity()).accountId,
                            uaProfileTagName);
                    return true;
                });
    }
}

File: app/src/main/java/xyz/klinker/messenger/fragment/settings/MmsConfigurationFragment.kt
package xyz.klinker.messenger.fragment.settings

import android.Manifest
import android.app.AlertDialog
import android.content.Intent
import android.content.pm.PackageManager
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.support.v7.app.AppCompatActivity
import android.util.Log

import com.codekidlabs.storagechooser.StorageChooser

import xyz.klinker.messenger.R
import xyz.klinker.messenger.api.implementation.Account
import xyz.klinker.messenger.api.implementation.ApiUtils
import xyz.klinker.messenger.shared.data.MmsSettings
import xyz.klinker.messenger.shared.data.Settings

class MmsConfigurationFragment : MaterialPreferenceFragment() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        addPreferencesFromResource(R.xml.settings_mms)
        initConvertToMMS()
        initMaxImageSize()
        initGroupMMS()
        initAutoSaveMedia()
        initDownloadLocation()

        // MMS APN settings
        initOverrideSystemSettings()
        initMmsc()
        initProxy()
        initPort()
        initUserAgent()
        initUserAgentProfileUrl()
        initUserAgentProfileTagName()
    }

    override fun onStop() {
        super.onStop()
        Settings.forceUpdate(activity)
    }

    private fun initConvertToMMS() {
        findPreference(getString(R.string.pref_convert_to_mms))
                .setOnPreferenceChangeListener { _, o ->
                    val convert = o as String
                    ApiUtils.updateConvertToMMS(Account.accountId,
                            convert)
                    true
                }
    }

    private fun initMaxImageSize() {
        findPreference(getString(R.string.pref_mms_size))
                .setOnPreferenceChangeListener { _, o ->
                    val size = o as String
                    ApiUtils.updateMmsSize(Account.accountId,
                            size)
                    true
                }
    }

    private fun initGroupMMS() {
        findPreference(getString(R.string.pref_group_mms))
                .setOnPreferenceChangeListener { _, o ->
                    val group = o as Boolean
                    ApiUtils.updateGroupMMS(Account.accountId,
                            group)
                    true
                }
    }

    private fun initAutoSaveMedia() {
        findPreference(getString(R.string.pref_auto_save_media))
                .setOnPreferenceChangeListener { _, o ->
                    val save = o as Boolean
                    ApiUtils.updateAutoSaveMedia(Account.accountId,
                            save)
                    true
                }
    }

    private fun initDownloadLocation() {
        findPreference(getString(R.string.pref_mms_save_location))
                .setOnPreferenceClickListener { _ ->

                    if (ContextCompat.checkSelfPermission(activity, Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_GRANTED) {
                        showStorageChooser()
                    } else {
                        if (activity != null && Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) {
                            activity.requestPermissions(arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), 10)
                        } else {
                            showStorageChooser()
                        }
                    }

                    false
                }
    }

    private fun showStorageChooser() {
        val chooser = StorageChooser.Builder()
                .withActivity(activity)
                .withFragmentManager((activity as AppCompatActivity).fragmentManager)
                .allowCustomPath(true)
                .setType(StorageChooser.DIRECTORY_CHOOSER)
                .actionSave(true)
                .skipOverview(true)
                .withPreference(MmsSettings.getSharedPrefs(activity))
                .build()

        chooser.show()
        chooser.setOnSelectListener { _ -> }
    }

    private fun initOverrideSystemSettings() {
        findPreference(getString(R.string.pref_override_system_apn))
                .setOnPreferenceChangeListener { _, o ->
                    val override = o as Boolean
                    ApiUtils.updateOverrideSystemApn(Account.accountId,
                            override)

                    if (override) {
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M && !android.provider.Settings.System.canWrite(activity)) {
                            AlertDialog.Builder(activity)
                                    .setMessage(com.klinker.android.send_message.R.string.write_settings_permission)
                                    .setPositiveButton(com.klinker.android.send_message.R.string.ok) { _, _ ->
                                        val intent = Intent(android.provider.Settings.ACTION_MANAGE_WRITE_SETTINGS)
                                        intent.data = Uri.parse("package:" + activity.packageName)
                                        intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK)

                                        try {
                                            startActivity(intent)
                                        } catch (e: Exception) {
                                            Log.e("MainActivity", "error starting permission intent", e)
                                        }
                                    }
                                    .show()
                        }
                    }

                    true
                }
    }

    private fun initMmsc() {
        findPreference(getString(R.string.pref_mmsc_url)).setOnPreferenceChangeListener { _, o ->
            val mmsc = o as String
            ApiUtils.updateMmscUrl(Account.accountId,
                    mmsc)
            true
        }
    }

    private fun initProxy() {
        findPreference(getString(R.string.pref_mms_proxy)).setOnPreferenceChangeListener { _, o ->
            val proxy = o as String
            ApiUtils.updateMmsProxy(Account.accountId,
                    proxy)
            true
        }
    }

    private fun initPort() {
        findPreference(getString(R.string.pref_mms_port)).setOnPreferenceChangeListener { _, o ->
            val port = o as String
            ApiUtils.updateMmsPort(Account.accountId,
                    port)
            true
        }
    }

    private fun initUserAgent() {
        findPreference(getString(R.string.pref_user_agent)).setOnPreferenceChangeListener { _, o ->
            val userAgent = o as String
            ApiUtils.updateUserAgent(Account.accountId,
                    userAgent)
            true
        }
    }

    private fun initUserAgentProfileUrl() {
        findPreference(getString(R.string.pref_user_agent_profile_url)).setOnPreferenceChangeListener { _, o ->
            val uaProfileUrl = o as String
            ApiUtils.updateUserAgentProfileUrl(Account.accountId,
                    uaProfileUrl)
            true
        }
    }

    private fun initUserAgentProfileTagName() {
        findPreference(getString(R.string.pref_user_agent_profile_tag)).setOnPreferenceChangeListener { _, o ->
            val uaProfileTagName = o as String
            ApiUtils.updateUserAgentProfileTagName(Account.accountId,
                    uaProfileTagName)
            true
        }
    }
}
